[포커게임 앱-Step6] 미션 제출합니다.

ehgud0670 opened the pull request:
> 궁금한 점

```swift
   private static func checkFourCardAndUpdateCards(cards: [Card]) -> ([Card],Card.Number)? {
        var cards = cards
        if let number = Rank.Combination.isFourCard(cards: cards) {
            cards.removeAll { (card) -> Bool in
                return card.number == number
            }
            return (cards,number)
        }
        return nil
    }
```

위의 코드처럼 반환 값이 두개가 필요한 것 같아 두개를 반환했는데 이렇게 구현해도 되는지 모르겠습니다. :sweat_smile:

ehgud0670 committed:
Refactor : ParticipantsNum 구조체 지우고 Participants 객체에 통합함
	이유1 : segmentedControl의 players 버튼을 누른 순간
	paritipants 객체가 결정된다고 생각하기 때문에 PokerGame 객체에게
	매개변수로 participants를 넘겨주기 위해서 기능 수정함

	이유2 : 이전의 코드는 ParticipantsNum구조체와 Participants객체가
	분리된 구조라 이상적이지 않다고 판단해 하나로 통합함.
	예를 들어 dealer가 2명있는 구조로 바뀌면 ParticipantsNum 구조체와
	Participants 객체의 코드를 모두 바꿔야 함으로 신경쓸 것이 많지만
	이를 하나로 통합하면Participants만 바꾸면 되므로 변화가 최소화된다.

ehgud0670 committed:
Refactor : 프로덕션 코드 변경사항에 맞게 테스트 코드 수정 및 불필요한 주석 제거

ehgud0670 committed:
Deleted : 카드덱에러인 DeckError를 의미없다고 판단해 삭제

Style: 개행 정리
       기준 1 : 프로퍼티 사이의 개행 없게 정리함
       기준 2 : class, struct, extension 의 경계에는 개행을 함

ehgud0670 committed:
Style: 클로져 반환 타입에 Void 라고 명시함
	이유 : 반환 타입이 Void인것을 명시적으로 보여주고 싶어서

ehgud0670 committed:
Style: 클로져 반환 타입에 Void 라고 명시함
            이유 : 반환 타입이 Void인것을 명시적으로 보여주고 싶어서

ehgud0670 committed:
Refactor: segmentedControl 선택했을때 enum 추출하는 부분에서 guard 문 적용
              * Controller 클래스 만들어서 index 값이 적절한지 판단하는 메소드문 구현

              * guard 문 적용한 이유
                : 해당 함수들은 리턴값이 옵셔널이고 배열의 index를 다룬다.
                  따라서 배열의 index가 적절치 못하면 nil이 반환될 수 있기에 guard문으로
                    가독성을 높였다.
              * Controller 만들어서 index 값이 적절한지 판단하는 메소드 만든 이유
                : DRY(Don't Repeat Yourself)원칙을 고수하기 위해

ehgud0670 committed:
Refactor: 뷰를 업데이트하는 메소드에 index 검증하는 코드 적용
	이유 : 해당 메소드들에 옵셔널!이 있다.
	       따라서 에러의 위험이 있기 때문에 검증하는 코드를 넣었다.

	* 개발자가 아무리 조심해도 프로그램은 버그가 생길 수 있다.
	  따라서 최대한 런타임 에러가 안나도록 코드를 작성할 필요가 있다.

ehgud0670 committed:
Refactor: 카드이미지 파라미턴의 값을 non-optional로 수정
	이유 : 카드 이미지가 있는지 없는지는 함수 외부에서 결정해야 되고
	함수 내부에서 검증할 필요는 없다고 생각하기 때문이다.

ehgud0670 committed:
Feat: 카드 클래스의 카드 총 개수를 나타내는 프로퍼티 추가
	이유 : CardsStackView의 setNotCardsViewHidden 메소드가
	외부의 cardsIndex에 의존하지 않게 하려고

ehgud0670 committed:
Style: 쓰임새를 더 잘 표현하기 위해 함수 이름 수정

ehgud0670 committed:
Refactor: 가독성을 위해 함수를 분리하고 쓰임새를 잘 나타내기 위해
	  메소드 이름 수정

ehgud0670 committed:
Refactor: 가독성을 위해 메소드 분리하고 적절한 이름으로 메소드 이름 수정

ehgud0670 committed:
Refactor: 가독성을 위해 if 문을 guard문으로 수정

ehgud0670 committed:
Refactor: 가독성을 위해 메소드 분리함
	* 메소드에서 사용할 private 한 연산 프로퍼티 생성

ehgud0670 committed:
Refactor: 메소드의 코드 수를 줄이기 위해 guard 문 합침
	* guard문의 목적도 일치하므로 합침

ehgud0670 committed:
Refactor: 메소드의 코드 수를 줄이기 위해 함수의 빈칸 삭제

ehgud0670 committed:
Refactor: 가독성을 위해 메소드 분리함

ehgud0670 committed:
Refactor : setConstraint에서 constant 값 재설정함
	 이유 : 요구사항의 출력화면처럼 보이기 위해

ehgud0670 committed:
Add: crown 이미지 추가

ehgud0670 committed:
Naming: configure 보단 setup이 적절할 것 같아 메소드 이름 수정

ehgud0670 committed:
Feat: ParticipantsStackView의cardsStackView 프로퍼티 추가 및 적용
	이유:  스택에 들어갈 뷰의 개수도 2개뿐이고
	       무엇이 있어야 하는지 명확해서,
	       또 cardsStackView가 class 타입이라서
	       index를 통해 update하는 것보다 프로퍼티를 참조해서
	       update 하는 걸로 결정함

ehgud0670 committed:
Docs: step6 관련하여 구현 순서 내용 작성

ehgud0670 committed:
Feat: 게임 조합 순위를 담은 Rank enum 생성

ehgud0670 committed:
Feat: 포카드인지 확인하는 메소드 추가
      * TDD를 도입하려고 했으나 잊고 프로덕션 코드부터 작성했습니다...
	다음 메소드부턴 TDD 방식을 도입하겠습니다!

ehgud0670 committed:
Refactor: 테스트 코드에 따라 프로덕션 코드 수정함
Feat: 테스트 코드 기능 구현함

     역시 테스트 코드...

ehgud0670 committed:
Feat: 랭크배열을 업데이트하는 함수에 FourCard 의 경우를 적용함
	* FourCard인 경우 랭크 배열에 FourCard를 추가하고
	  이외의 조합 검사를 위해 FourCard인 카드들
	  기존 카드 배열에서 삭제함

Refactor: isFourCard의 리턴값 Bool에서 Int로 변경
 	* FourCard에 해당하는 num값 반환하기 위해서

ehgud0670 committed:
Feat: 스트레이트인지 확인하는 메소드 구현 및 적용

ehgud0670 committed:
Feat: 트리플인지 확인하는 메소드 구현 및 적용

ehgud0670 committed:
Refactor : 재귀함수인걸 감안하여 else if문으로 수정

ehgud0670 committed:
Refactor: 정렬된 키&값 배열을 반환하도록 코드 수정

ehgud0670 committed:
Feat: 투페어인지 확인하는 메소드 추가 및 적용

ehgud0670 committed:
Refactor: 재귀함수가 불필요하다고 판단하고 메소드 코드 수정

ehgud0670 committed:
Feat: 원페이인지 검사하는 메소드 구현 및 적용

ehgud0670 committed:
Style: enum의 이름이 Combination이 더 맞는것 같아 수정

ehgud0670 committed:
Feat: Rank라는 구조체 만들고 Combinaiton 이동

ehgud0670 committed:
Refactor: 조합 업데이트하는 함수들만 확장해서 분리함
	이유: 1. 확장에 열려야하는 기준으로 전에 했어야 하는 걸
	      지금 하게됨
	      2. 기존의 Participant의 기능과
		 분별되어 보이게 하기 위해서

ehgud0670 committed:
Refactor: Combination 배열에서 Rank 배열로 바꿈
	이유: 같은 핸드를 비교하려면 가장 높은 수를 비교해야 하기 때문에
	가장 높은 수를 고려하여 코드 수정함

	애초에 이전의 함수들은 조합에서 가장 큰수를 반환하게 구현해놨음

ehgud0670 committed:
Feat : 조합이 아닌 카드의 숫자가 비교할때 필요한 경우와
	조합이 하나도 없을때를 대비하여 isOneCard 메소드를 구현함

ehgud0670 committed:
Refactor: 이후에 OneCard를 찾을 것을 대비해 투페어 확인하는 메소드에서
          남은 카드를 반환하도록 메소드 코드 수정

ehgud0670 committed:
Refactor: 원카드인 경우를 대비하여 원페이 체클하는 함수가 남은
	카드를 반환하도록 코드 수정함

ehgud0670 committed:
Feat: 원카드 랭크 결정하는 함수 생성 및 적용

ehgud0670 committed:
Refactor: 랭크를 결정하는 대부분의 코드를 랭크로 옮김
	  * 반환값이 두 개(cards: [Card],ranks: [Rank])가 되도록 설정함
	  * static 한 함수로 만듦

ehgud0670 committed:
Feat: 플레이어들에게 카드를 다 나눠준뒤에 플레이어들이 랭크를 업데이트하도록
      기능 구현함

ehgud0670 committed:
Style: 철자가 틀려서 메소드 이름 변경

ehgud0670 committed:
Feat: Ranks을 탐색할 수 있도록 확장해서 메소드 구현

ehgud0670 committed:
Feat: 1.Rank에 Comparable 적용 (확장 x => Rank라 기존에 있어야 할 기능이라서)
      2.Pariticipant에 Comparable 적용 ( 확장 o => 기존에 없는 기능이라서)

      * Rank에 Comparable 적용한 논리
	1. Rank의 Combination을 먼저 비교하고 오른쪽이 더 크다면 true을 반환한다.
	2. 만약 두 Rank의 Combination이 같다면 number을 비교해 오른쪽이 더 크다면 true을 반환한다.

      * Participant에 == (Equatable)을 적용한 논리
	1. 두 rank 배열의 길이가 같지 않다면 false 반환
	2. 두 rank 배열의 같은 index의 원소가 같지 않다면 false 반환
	3. 1,2에 해당하지 않는다면 true 반환

      * Participant에 <(Comparable)을 적용한 논리
        1. 두 rank 배열 길이중 작은 것을 min으로 결정함
	2. min의 길이만큼 두 rank 배열의 같은 index의 원소끼리 비교하는데
	   오른쪽 배열의 원소의 combination이 더 크다면 true을 반환
	3. 2에 해당하지 않는 경우 min의 길이에 해당하지 않는 원소로 판단한다.
	   * 왼쪽 배열의 길이가 min보다 크다면 false를 반환
 	   * 오른쪽 배열의 길이가 min보다 크다면 true를 반환
	4. 3에 해당하지 않는 경우 min의 길이만큼
	   두 rank 배열의 같은 index의 원소의 number를 비교하고
	   오른쪽 배열의 원소의 number가 더 크다면 true을 반환한다.
 	5. 4에 해당하지 않는 경우에는 같은 것이므로 false를 반환한다.

ehgud0670 committed:
Feat: 승자 결정하는 기능 구현
	* 승자를 나타내는 winner 프로퍼티 추가

ehgud0670 committed:
Refactor: 반환 값을 Card.Number로 변경

	이유: Rank배열을 건네주고 다시 받는 방식은 좋지 않는 표현이다.
	반환된 number를 가지고 직접 프로퍼티인 Rank 배열에 원소를
	추가하는 것이 더 적절한 방법이다.

ehgud0670 committed:
Refactor: 반환값을 non-optional로 처리
	이유 : 원래는 isOneCard에서 카드배열을 반환할 필요가 없어
	타입을 optional로 뒀지만, 이러면 그 외 나머지 반환값을 모두 번거롭게 처리해야
	해서 non-optional로 바꿈
	* oneCard가 반환할때는 빈 카드배열을 반환하는 걸로 처리함

ehgud0670 committed:
Style: OverlappedCardsStackView가 더 적절한 이름 같아 함수명 변경함

ehgud0670 committed:
Style: 매개변수 이름이 combination이 더 맞는 것 같아 매개변수 명 수정함

ehgud0670 committed:
Feat: 카드들과 왕관 이미지를 넣을 스택뷰 생성

ehgud0670 committed:
Style: 뷰를 더하는 더 적절한 의미로 메소드 명 변경

ehgud0670 committed:
Refactor: CardsAndCrownStackView가 연결되게끔 코드 수정

ehgud0670 committed:
Refactor: 뒤늦게 extension으로 분리안된 것을 보고 분리함

ehgud0670 committed:
Refacotor: winner인 것을 Bool타입인 프로퍼티로 선언하여 코드에 적용함
	   이유: 타입이 Participant인 winner 프로퍼티를 따로 두면
	   번거로워질것 코드 수정함

ehgud0670 committed:
Feat: 승자인 플레이어의 카드들 오른쪽에 왕관이 보이도록 기능 구현
	* 왕관 이미지 작은 걸로 새로 추가함

ehgud0670 committed:
Docs: step6에 맞게 파일 작성
	* 구현사항의 내용을 보강함
	* 출력화면을 기입함

ehgud0670 committed:
Refactor : enum의 rawValue을 0부터 오름차순으로 되게 코드 수정함
	이유 : 생각해보니 이전과 같은 rawValue 는 필요없다고 생각함
		=> 승자를 비교할 때 조합이 더 크면 무조건 이긴식으로
		   코드를 작성했기 때문에 이전과 같은 rawValue는 필요없다.

ehgud0670 committed:
Refactor: 랭크를 업데이트하는 메소드 모두 Rank에 두기로 함
  	이유 : 1. 생각해보니 랭크를 체크하는 것은 랭크 구조체만 해야 한다고 판단함(한가지 책임만 가져야 해서)
	       2. 코드가 쓸데없이 복잡하다고 생각함

ehgud0670 commented:
방금 또 업데이트 했습니다~!

ehgud0670 committed:
Style: at 매개변수 레이블 추가함

ehgud0670 committed:
Style: 메소드의 위치 변경
      이유 : 호출한 순서대로 메소드를 위치하기 위해

ehgud0670 committed:
Docs: 더 나은 설명이 되기 위해 수정함

ehgud0670 committed:
Refactor: max()을 사용하도록 코드 변경
	* 이유1: 본래 목적은 가장 큰 객체를 리턴만 하면 되기 때문이다,
		 반면 이전 코드의 sorted는 불필요하게 배열을 리턴한다.
	* 이유2: sorted을 쓰는 것보다 코드가 간결해지기 때문에

ehgud0670 committed:
Refactor: view들의 autolayout 재설정함
	* safeLayoutGuide의 기준에 맞게 재설정함
	* 중심을 기준으로 레이아웃 잡을 수 있는 것들은 leadingAnchor,
	  trailingAnchor 설정을 지우고 centerX나 centerY로 레이아웃 재설정함

godrm commented:
`private static func checkFourCardAndUpdateCards(cards: [Card]) -> ([Card],Card.Number)?` 
이것만 보면 함수가 두 가지 일을 하는 것처럼 보이네요. 하나만 하는 함수 두 개가 있으면 안 될까요?
꼭 매개변수를 바꿔서 리턴해야 하는건가요? 리턴타입이 튜플과 배열의 조합인데 옵셔널이라 상당히 복잡하네요. 
복잡하다는 것은 나쁜 냄새가 나는 겁니다. 본인도 그걸 느끼고 있구요. 
개선할 부분은 코드에서 보고 맥락을 보고 다른 피드백이 있으면 남길게요.

godrm commented on CardGameApp/CardGameApp/Controller/Controller.swift:
```diff
@@ -0,0 +1,17 @@
+    static func verifyIndex(at index: Int, arrLen: Int) -> Bool {
+        return index < arrLen
+    }
+    
```
함수가 딱 하나 뿐이네요. 어떤 의도가 있는 걸까요? 어디에 사용하는지 전혀 맥락을 모르겠습니다. 
그냥 비교만 하는거면 그대로 써도 되지 않을까요?

ehgud0670 replied:
맞습니다. guard문으로 해당 index와 배열의 길이를 비교하는 코드를 다른 메소드에서도 하다보니
이걸 재사용하고 싶어서 만들어 놓긴 했습니다.

1. 근데 생각해보니 단순히 비교만 하는 거라 함수를 호출할 필요는 없을 것 같고 
2. Controller의 원래 목적이 수업에 들은 것처럼 들어온 input 값이 맞는 것인지 (verify) 확인하는 용도라고
    배웠기 때문에 위 메소드는 이 목적이 맞지 않으므로 없애는 게 나을 것 같습니다. 

수정하겠습니다~

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -37,21 +40,25 @@ struct Card {
     
     private let suit: Suit
-    private let number: Number
+    let number: Number
```
왜 갑자기 number가 private 에서  public으로 풀렸을까요? 값을 가져가지 않고 비교할 방법은 없었을까요?

ehgud0670 replied:
네, number 프로퍼티는 private으로 두고,  카드들끼리 비교할 수 있는 구조로 만들어보겠습니다. 
어차피 suit는 순위에 영향을 주지않고 number만 영향을 주므로 Card의 Comparable을 number끼리만 비교하는 코드로 작성하겠습니다.
(로직을 하위모듈에서 구현하고 값만 return하는 형식이 가능하면 getter,setter를 쓰지 않도록 신경을 쓰겠습니다. 💪)

godrm commented on CardGameApp/CardGameApp/Model/Participant.swift:
```diff
@@ -8,11 +8,29 @@
+        return self.rawValue + plus == other.rawValue
+    }
+    
+    func isEqual(minus: Int, other: Card.Number) -> Bool {
```
위에 `isEqual(plus: other:)` 와 타입은 동일한 데 `isEqual(distance: other:)`로 하고 +, -를 넣는 방식은 어떤가요?

ehgud0670 replied:
오 넵 감사합니다. 반영하도록 하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Participant.swift:
```diff
@@ -8,11 +8,29 @@
+    var cardsCount: Int {
+        return cards.count
+    }
+    var ranks = [Rank]()
```
rank는 public으로 해서 외부에서 값을 바꿀 수 있는 값인 건가요?

ehgud0670 replied:
사실 외부에서 사용하지 않는데 private으로 두지 않았습니다. 
반영토록 하고 다음부터는 코드를 더 주의깊게 보겠습니다~!

godrm commented on CardGameApp/CardGameApp/Model/Participant.swift:
```diff
@@ -25,7 +43,23 @@ class Participant: CardSearchable {
+    private(set) var isWinner = false
+    func setWinner() {
+        isWinner = true
+    }
```
isWinner 가 Participant 마다 Bool 타입으로 있는 것보다는 
상위 모듈 Participants 에 `private var winner : Participant?` 이렇게 있는 게 어떨까요?

ehgud0670 replied:
사실 뷰 업데이트를 할때  winner 인 participant만 왕관이미지를 띄우는 것을 `isWinner`로  구분하여 하려고 했던 것인데 위에 방법으로 한번 해보겠습니다.  
제 방식에서 `isWinner`는 프로퍼티인데 외부에서 노출이 되서 좋지 않은 방식인가요?? 궁금합니다. 🤷‍♂️

ehgud0670 replied:
아 `private var winner`에서 말뜻을 알아들었습니다. winner를 결정하는 것을 Participants만 할수 있게 되는 군요. 무릎을 탁치고 갑니다.. 👍

godrm commented on CardGameApp/CardGameApp/Model/Participant.swift:
```diff
@@ -25,7 +43,23 @@ class Participant: CardSearchable {
+    
+    func updateRanks() {
+        ranks = Rank.checkCombiAndGenerateRanks(with: cards)
+    }
```
이 부분은 왜 분리해서 구현한건가요? 프로토콜이 있는 것도 아닌데 나눈 이유가 궁금합니다.

ehgud0670 replied:
음 랭크 배열이라는 새로운 프로퍼티가 생겼고, 이에 대한 업데이트 하는 함수도 새로 만들어졌으니깐
확장한다는 의미로 extension해서 해당 메소드를 만들었습니다. 
혹시 프로토콜을 채택해야지만 분리를 해야 할까요.. 궁금합니다. 🤷‍♂️

godrm commented on CardGameApp/CardGameApp/Model/Participants.swift:
```diff
@@ -9,21 +9,33 @@
+            }
+        }
+    }
+    let playersNum: PlayersNum
```
PlayersNum 타입이름이나 playersNum 변수 이름에서 `Num`을 붙이는 것보다는 구체적인 역할을 명시하는 게 좋겠습니다.
스위프트 API 디자인 가이드에서도 단어를 줄여쓰기 보다는 좀 길어지더라도 역할에 대한 단어를 명확하게 쓰길 권장하고 있습니다.

ehgud0670 replied:
명심하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Participants.swift:
```diff
@@ -35,3 +47,40 @@ extension Participants {
+    
+}
+
+extension Participants.PlayersNum: CustomStringConvertible {
```
Participants.PlayersNum 을 확장했다가 Participants 확장하고 다시 Participants.PlayersNum 확장하는 코드가 뒤섞여 있어서 읽기 어렵네요. 같은 타입끼리 묶어주거나 같은 파일로 분리하는 것이 어떨까요?

ehgud0670 replied:
넵! 반영하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -8,24 +8,90 @@
+        return true
+    }
+    
+}
```
Participant를 확장하는 것은 왜 PokerGame에 있는걸까요? 
여기서만 써야 하더라도 특정 타입에 대한 확장은 별도 파일로 분리하는 게 좋겠습니다. 
타입 정보가 여기저기 흩어져서 찾아야 하고 어렵습니다. 타입/파일을 조합해서 읽기 편하도록 개선해보세요.

ehgud0670 replied:
네, 읽기 편하도록 타입/파일로 조합하도록 하겠습니다~ 👍

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -53,21 +119,46 @@ class PokerGame {
+    private func updateParticipantRanks() {
+        participants.searchParticipants { (participant) -> (Void) in
+            participant.updateRanks()
+        }
```
이 정도는 그냥 아래 determineWinner() 에서 `participants.determineWinner()` 메소드 하나로 해도 될 것 같습니다.  
게임에서 처리하려는 기능은 우승자를 찾는거죠. 
내부 메소드 구현을 항상 상위에서 클로저로 모두 다 구현해야 한다는 의미는 아닙니다. 
상위에서 구현해야 하는 경우도 있을 수 있고, 하위 모듈에서 복잡하게 구현하고 그걸 메소드로 호출해서 가져다 쓰기만 하면 되는 경우도 있는 거죠.
우승자를 찾는 것은 participants 에게 모두 들어 있어도 될 것 같습니다.

ehgud0670 replied:
네 , 로직을 하위 모듈에서 구현하도록 하겠습니다. 지금 생각해보니 Participants가 해야겠네요. 감사합니다.

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -53,21 +119,46 @@ class PokerGame {
             participantsCount += 1
         }
-        return deck.count >= stutNum * participantsCount
+        return participantsCount
```
이렇게 값 하나 가져오는 것 조차 클로저로 상위에서 구현해야 하는 것은 아닙니다. 
이런건 그냥 participants.count 라고 호출하는 게 더 간단하니까요.

ehgud0670 replied:
네 반영하도록 하겠습니다. 👍

godrm commented on CardGameApp/CardGameApp/Model/Rank.swift:
```diff
@@ -0,0 +1,317 @@
+                    return num.key
+                }
+            }
+            return nil
```
이 메소드는 [Card] 배열이나 Rank 에 있는게 어떨까요? 
Rank가 인스턴스가 없이 계산만 하는 static 클래스라면 매개변수로 [Card]를 매번 전달해야 겠네요. 
그래서 질문했던 구조가 된 것 같습니다.

`Rank.init(with: [Card])` 형태가 되고, Rank가 [Card]을 갖고 있으면 여기 메소드들은 
[Card]를 모두 참조하니까 응집도가 높은 클래스가 될 것 같습니다.
static으로 하려니까 구조가 이상해진 것 같습니다.

ehgud0670 replied:
넵 Ranks라는 객체가 프로퍼티로 [Rank]를 가지고 있고 내부에서 랭크를 업데이트 하는 메소드를 만들어보겠습니다. 피드백 감사합니다. 👍

ehgud0670 replied:
네, static은 객체지향적이지 않은 방식인데 제가 별 생각없이 짠 것 같습니다. 피드백 반영하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Rank.swift:
```diff
@@ -0,0 +1,317 @@
+            let nums = generateNums(cards: cards)
+            guard nums.count >= fiveCardsCount else {
+                return nil
+            }
```
이 앞부분이 비교하는 메소드들 마다 비슷하게 반복되는 것 같네요. 일반화할 수 없을까요?
매번 generateNums()를 호출해야만 할까요?
generateNums()가 넘겨주는 구조가 `Dictionary[Card : Int]` 형태면 어떨까요?

ehgud0670 commented:
맞습니다. 좀 지저분한 것 같습니다..
우선 두개의 반환값 [Card]와 Card.Number을 설명드리자면, 
* [Card] 배열은 제가 해당 카드들 중에 하나의 조합이 있다면 검사를 끝내는 게 아니라 이후에 더 조합이 있는 경우를 찾기 위해서  남은 카드를 리턴하기 위한 목적이고,

* Card.Number는 해당 조합이 맞는 경우 checkCombiAndGenerateRanks 메소드에서 Rank 객체를 만들때 
해당 조합의 숫자(Card.Number)를 집어넣기 위한 목적입니다. 

> 해결책 

그런데 JK의 피드백을 받고 생각해보니 해당 조합을 제거하고 나머지 카드만 배열에 남게하는 메소드를  따로 만들 수 있을  것 같습니다. 해당 조합이 맞는 경우에 checkCombiAndGenerateRanks에서 카드배열을 업데이트 하는 구조로 만들어보겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/Rank.swift:
```diff
@@ -0,0 +1,317 @@
+                isFifth = true
+            } else {
+                newCards.append(card)
+            }
```
이 함수들이 뭘 하려는건가요. combination으로 선택한 카드를 빼려는건가요?
만약 그런거면 원래 카드 배열에서 combination 카드 배열만 빼도 되지 않을까요? 
number로 비교하려다보니 너무 복잡해졌습니다.

ehgud0670 replied:
반영하였습니다.

godrm commented on CardGameApp/CardGameApp/View/CardsAndCrownStackView.swift:
```diff
@@ -0,0 +1,70 @@
+        addArrangedSubview(overlappedCardsStackView)
+    }
+    
+    private func addAndSetConstraintForCrownImageView() {
```
함수 이름에 And가 들어가는 것은 좋지 않습니다. 이미 여러 작업을 하고 있다는 의미가 되니까요. 
add를 붙이지 않아도 될 것 같네요.

ehgud0670 replied:
네 알겠습니다. And를 사용하지 않도록 하겠습니다.

ehgud0670 replied:
`CardsAndCrownStackView` 라는 클래스도 만들어놨는데 클래스 이름에는 And가 들어가도 괜찮을까요?? 🤷‍♂️

godrm replied:
그것도 이상하네요

godrm reviewed (commented):
점수 계산하는 부분에서 static 클래스 메소드로 구현하다보니 어색해진 부분이 많네요.
이렇게 매개변수로 사용하는 게 많으면 인스턴스로 만들려고 노력해보세요. 
자세한 내용은 피드백을 확인해보시고, 이해가 안되는 부분은 질문하세요.

ehgud0670 committed:
Deleted: Controller 클래스 지우고 인덱스와 배열 길이를 비교하는 걸
	 각 함수에서 직접하도록 코드 수정함

	 이유: 인덱스와 배열 길이를 비교하는 것은 코드가 단 한 줄이면 되므로
	       굳이 함수를 만들 필요는 없다고 생각함

ehgud0670 committed:
Refactor: isEqual을 하나로 합침
	 대신 distance에 +,- 값이 모두 들어오는 형태로 만듬

ehgud0670 committed:
Style: 타입/파일로 조합하도록 분리된 확장들 파일이 있는 곳으로 옮김

ehgud0670 committed:
Refactor: 외부에서 접근이 불가능하도록 private 접근제한자를 둠

ehgud0670 committed:
Style: 1. 함수 이름에 And가 들어가는 것은 좋지 않으므로 함수 명 변경
       2. setup 메소드에는 setup 이름의 메소드들만 들어가도록 하고
	  setup 이름의 메소드에서 add 함수나 setConstraint 함수를 두도록
	  함수들을 분리함

ehgud0670 committed:
Style: 클래스 이름에 And를 넣는 것은 좋지 않으므로 이름 변경
       이유: 하나의 모듈이나 함수는 하나의 목적만을 가져야 하는데
	    And라 두면 여러가지 일을 하는 것처럼 보이기에
	    네이밍에 And를 두는 것은 좋지 못한 표현이다.

ehgud0670 committed:
Refactor: 비교하는 것을 Card.Number가 아닌 Card로 둠
	 이유: Card.Number는 프로퍼티로 private 하게 두는 게 제일 좋다.
	       따라서 노출이 되는 Card끼리만 Comparable을 사용해서 비교되도록 하였다.
	 헤멘 점 : Card를 디셔넉리의 key로서 사용하려고 Hashable을 채택하니깐
		   Duplicate한 key가 계속 발견되는 런타임에러가 났다.
		   hash함수를 구현하니 해결되었는데, 앞으로
		   구조체된 클래스든 hashable을 채택하면 hash 함수를 사용해야겠다.

ehgud0670 committed:
Refactor: 배열의 길이가 적당한 길이인지 확인하는 메소드 구현 및 적용
	 * 비슷한 코드를 함수를 만들어서 재사용하게 함
	  Dont' Repeat Yourself

ehgud0670 committed:
Style: 읽기 편하도록 타입끼리 묶어 놓음

ehgud0670 committed:
Style: 폴더 Model에서 Models로, View에서 Views로 변경함

ehgud0670 committed:
Style: 더 정확한 설명을 위해 PlayersNum에서 PlayersNumber로 타입 이름 수정함
	변수들도 playersNum에서 playersNumber로 변수 이름 수정함

ehgud0670 committed:
Refactor: 승자를 결정하는 역할은 Participants 가 하는게 더 맞는 표현이므로
	  메소드를 PokerGame에서 Particiants로 둠

	* 하위 모듈에서 할수 있는 것은 하위 모듈에서 하자.

ehgud0670 committed:
Refactor: private 한 winner(type: Participant) 프로퍼티을 생성 및 적용

	이전의 코드의 단점: participant의 winner 프로퍼티에 접근해서
	임의로 승자로 만들 수 있다.

	지금 코드의 장점: private으로 두었기 때문에 winner를 결정하는 것은
	Participants만 가능하기 때문에 외부에서 임의로 승자를 만들 수 없다.
	또 Partipant는 class타입이기 때문에 ===을 이용해 참조 동일성으로
	승자를 나타낼 수 있다.

ehgud0670 committed:
Refactor: Deck.count처럼 참여자들의 수를 나타내는 count 연산 프로퍼티 생성 및 적용

	 이유 : 값 하나 가져오는 것을 클로저로 구현하기 보단
	        단순히 get 하는 것이 더 간단하다.
		해당 count는 적절한 경우이고, 외부에서 get만 가능하도록
		연산 프로퍼티로 작성하였다.

ehgud0670 committed:
Refactor: 게임 리셋이 되었을 때 랭크도 리셋되게끔 코드 수정

ehgud0670 committed:
Refactor: 앱에 shake 이벤트를 주는 경우 새로운 게임 객체가 생성되는 것이
	아니라 기존의 게임으로 새로운 round을 시작하도록 함

        기존의 방식은 새로운 게임 객체가 매번 생성되는 구조이기 때문에
	카드가 남아있어도 사용할 수 없었다.
	때문에 shake하는 경우에는 남은 카드도 사용할 수 있도록 하였다.

	* ViewController 에 optional 타입인 game을 하나 두었다.

ehgud0670 committed:
Refactor: 랭크 배열을 가지고 있는 Ranks객체 생성 및 적용

ehgud0670 committed:
Refactor: == 함수를 명시적으로 구현함

ehgud0670 committed:
Refactor: combination의 static 메소드를 ranks의 instance 메소드로 옮김
	이유1: 결과적으로 반환값을 card배열만 가질수 있게됨
	이유2: 프로퍼티에 대해 인스턴스 메소드를 쓰므로 더 객체지향적인 방법이다.

ehgud0670 committed:
Style: result보다 updatedCards 가 더 적절한 이름 같아 변수명 바꿈

ehgud0670 committed:
Refactor: 메소드에 And 표현이 있어 하나의 메소드가 하나의 역할만을 하도록
	  메소드 분리함
	  1) isXXX : 해당 조합이 맞으면 맞는 카드중 제일 높은 숫자의 카드를 반환
	  2) removeXXX: 카드들 중에 해당 조합인 카드들 제거하고 남은 카드들을 반환

ehgud0670 committed:
Refactor: 함수이름에 And표현이 들어가 있으므로 수정하므로 메소드 분리함

ehgud0670 committed:
Refactor: 스트레이트를 제거하는 함수를 수정함
	기존의 방식은 넘버를 일일히 확인해 코드가 매우 길었지만
	card의 deepEqual함수를 사용하여 코드를 많이 줄일 수 있었다.

ehgud0670 committed:
Deleted: 기존 static 메소드들이 Ranks의 메소드가 됨에 따라
	기존의 테스트 코드 삭제함

ehgud0670 committed:
Refactor: 확장하려던 것을 하지 않고 코드를 위로 옮김

ehgud0670 committed:
Style: 함수명을 generateCards 가 더 적절해 수정함

ehgud0670 commented:
피드백 반영하였습니다. 

1. Controller 클래스 삭제
2. number는 private으로 두고 Card를 비교하게 함 
3. isEqual함수를 두개로 합침. isEqual(distance: other:)
4. Ranks라는 객체 만들고 private 한 프로퍼티로 둠
5. Participants 에서 winner(type: Participant?)를 프로퍼티로 두고 private하게 함
6. Participants에 참여수를 나타내는 count 연산 프로퍼티를 둠 
7. 기존의 static 한 메소드들을 인스턴스 메소드로 바꾸고 프로퍼티에 접근하게 함
8. 해당 카드배열이 충분한 길이를 가지고 있는지 확인하는 코드를 일반화하고, generateCards의 반환값을 [Card:Int]로 수정함 
9. 스트레이트를 제거하는 함수를 단순화함( Card의 isDeepEqual()을 사용함)
10. 함수명, 클래스명에 And가 들어간 것들 다 수정하고 역할도 분리함 

> 의문점 

사실 확장을 무슨 기준으로 해야하는지 정확히 모르겠습니다. 
프로토콜을 구현해야만 extension을 써야 하는 것인가요? 🤷‍♂️

godrm commented on CardGameApp/CardGameApp/Models/Participants.swift:
```diff
@@ -77,7 +88,15 @@ extension Participants {
+        return lhs.rawValue < rhs.rawValue
+    }
+    
+}
```
이 정도면 꼭 extension 으로 분리하지 않고 같이 구현해도 좋을 것 같아요. 참고하세요.

godrm commented on CardGameApp/CardGameApp/Models/Ranks.swift:
```diff
@@ -0,0 +1,273 @@
+        let sameNumberCards = generateCards(cards: cards)
+        return sameNumberCards.last!.key
+    }
+}
```
extension 이 선언부분보다 위에 나오는데 이것도 같이 구현하거나 아래에 있으면 좋겠습니다.
어떤 경우는 위에 있고, 어떤 경우는 아래 있고 일관성이 없네요. 잘 지켜주세요.

ehgud0670 replied:
넵, 확인해보니 정말 어지럽네요 ㅋㅋ 선언 부분 아래로 extension이 있도록 반영하였습니다. 
다음 부터는 일관성있게 extension이 위치하도록 하겠습니다. 감사합니다~!

godrm commented on CardGameApp/CardGameApp/Models/Ranks.swift:
```diff
@@ -0,0 +1,273 @@
+    }
+    
+    static func < (lhs: Ranks, rhs: Ranks) -> Bool {
+        
```
공백이나 한 줄 띄는 것도 의도를 표현하기를 바랍니다. 
선언 부분 앞뒤로 띄는 경우가 있는데 아닌 경우도 있는 것 같아요.

ehgud0670 replied:
네 선언 부분 앞뒤로 띄는 것 맞습니다. 근데 지금 확인해보니 아닌 경우가 꽤있네요.. 😂 반영하였습니다. 감사합니다.
